pkg/migration/migrationmanager/manager.go

```diff
@@ -234,3 +234,131 @@ func (m *Manager) Migrate(ctx context.Context, from, to clusterversion.ClusterVe
+			return err
+		}
+		var found bool
+		found, jobID, err = m.getRunningMigrationJob(ctx, txn, version)
+		if err != nil {
+			return err
+		}
+		if found {
+			return nil
+		}
+		var j *jobs.Job
+		j, err = m.jr.CreateJobWithTxn(ctx, migrationjob.NewRecord(version, user), txn)
+		if err != nil {
+			return err
+		}
+		jobID = *j.ID()
+		return nil
+	}); err != nil {
+		return false, 0, err
+	}
+	return alreadyCompleted, jobID, nil
+}
+
+func (m *Manager) getRunningMigrationJob(
+	ctx context.Context, txn *kv.Txn, version clusterversion.ClusterVersion,
+) (found bool, jobID int64, _ error) {
+	const query = `
+SELECT id, status
+	FROM (
+		SELECT id,
+		status,
+		crdb_internal.pb_to_json(
+			'cockroach.sql.jobs.jobspb.Payload',
+			payload,
+      false -- emit_defaults
+		) AS pl
+	FROM system.jobs
+  WHERE status IN ` + jobs.NonTerminalStatusTupleString + `
+	)
+	WHERE pl->'longRunningMigration'->'clusterVersion' = $1::JSON;`
+	jsonMsg, err := protoreflect.MessageToJSON(&version, false /* emitDefaults */)
+	if err != nil {
+		return false, 0, errors.Wrap(err, "failed to marshal version to JSON")
+	}
+	rows, err := m.ie.Query(ctx, "migration-manager-find-jobs", txn, query, jsonMsg.String())
+	if err != nil {
+		return false, 0, err
+	}
+	parseRow := func(row tree.Datums) (id int64, status jobs.Status) {
+		return int64(*row[0].(*tree.DInt)), jobs.Status(*row[1].(*tree.DString))
+	}
+	switch len(rows) {
+	case 0:
+		return false, 0, nil
+	case 1:
+		id, status := parseRow(rows[0])
+		log.Infof(ctx, "found existing migration job %d for version %v in status %s, waiting",
+			id, &version, status)
+		return true, id, nil
+	default:
+		var buf redact.StringBuilder
+		buf.Printf("found multiple non-terminal jobs for version %s: [", redact.Safe(&version))
+		for i, row := range rows {
+			if i > 0 {
+				buf.SafeString(", ")
+			}
+			id, status := parseRow(row)
+			buf.Printf("(%d, %s)", id, redact.Safe(status))
+		}
+		log.Errorf(ctx, "%s", buf)
+		return false, 0, errors.AssertionFailedf("%s", buf)
+	}
+}
+
+func (m *Manager) listBetween(
+	from clusterversion.ClusterVersion, to clusterversion.ClusterVersion,
+) []clusterversion.ClusterVersion {
+	if m.knobs.ListBetweenOverride != nil {
+		return m.knobs.ListBetweenOverride(from, to)
+	}
+	versions := clusterversion.ListBetween(from, to)
+	if len(versions) == 0 {
+		// We're attempt to migrate to something that's not defined in cluster
+		// versions. This only happens in tests, when we're exercising version
+		// upgrades over non-existent versions (like in the cluster_version
+		// logictest). These tests explicitly override the
+		// binary{,MinSupportedVersion} in order to work. End-user attempts to
+		// do something similar would be caught at the sql layer (also tested in
+		// the same logictest). We'll just explicitly append the target version
+		// here instead, so that we're able to actually migrate into it.
+		versions = []clusterversion.ClusterVersion{to}
+	}
```

Comment:
Is there a need for this special case now that we have a way to override what's returned through the testing knob?

Reply:
This is gone now.